Use the Grid for the one and two-column modal dialogs.

jeresig opened the pull request:
This diff adds in support for the one and two-column modal dialogs, using the new grid system. See #65 for the details of how the grid is ready.

A couple things that had fuzzy definitions in the design that I made a decision on:
* I added a 64px margin to the top and bottom of the modal, which seems to match the current design.
* I used a single 12-column grid for all two-column modals (which will be the same one I use for the standard modal, too).
* For the "11-column" modal the new design calls for a 128px space between the two pieces of text. I implemented that as: (in Sidebar: 32px space + 32px gutter + in Content: 64px space). This helps us to avoid having to implement a brand new type of grid just for this one modal.
* If the contents have a footer then I add a 48px padding to the bottom so that when the content scrolls up it looks nice (Zeplin shows it as 40px, but that's not one of our standard spacing sizes).

I haven't implemented mobile support yet, I'll handle that in a follow-up diff.

jeresig commented:
Deploy preview for *wonder-blocks* ready!

Built with commit e8ad0db004eeaaa5a053541bd39611a1f8023f9c

https://deploy-preview-62--wonder-blocks.netlify.com

jeresig commented:
Deploy preview for *wonder-blocks* ready!

Built with commit 0f2804cc1d27cc09dbeef810af62a17732536c1f

https://deploy-preview-62--wonder-blocks.netlify.com

matchu commented:
Eep, a bit surprising and scary to see so many changes to the modal baked into this one! 😬😬😬 Are these meant to be prototype-y to enable you to try out the grid, or are they meant to be finalized implementations of these layouts? That'll affect how carefully I review them.

jeresig commented:
@matchu I guess it depends on how you want to use what I implemented. I did try to implement all of the modal changes to the best of my ability and matching the spec as closely as possible - I do hope that we can use it! It was challenging as I had to make sure that the grid could work for every modal case we wanted to throw at it, which essentially meant implementing the entire two-column (and one-column) modal. If you want to just focus on the modal-related changes, that's fine, I can defer to @kevinbarabash for input on the Grid stuff. Up to you!

matchu commented on packages/wonder-blocks-modal/components/two-column-modal.js:
```diff
@@ -27,6 +25,15 @@ type Props = {
+    footer?: React.Node,
+
+    /** The main column's content. */
+    contents: React.Node,
```
I'm not sure this column actually is more "main" than the other. There's [some discussion][1] in the spec about this, and it's a bit in-flux. I proposed `leftColumn` and `rightColumn` to avoid the implied hierarchy, but then Kevin pointed out that maybe there _is_ some natural hierarchy because only the main column is present in the mobile design, then I pointed out that maybe that's not the intent of the design spec? We meant to ask Tabitha, but I'm not sure where we got with that. I've added a Zeplin comment now (51).

In any case, if we do decide to go with the hierarchical naming, I think this is `content` in the spec rather than `contents`.

[1]: https://docs.google.com/a/khanacademy.org/document/d/1__cz2bHvZGWtBIcf_W0IGpGtgu1HKgHIjS9mmYsjbxk/edit?disco=AAAABw4OXrQ

jeresig replied:
Hmm - ok! So when I saw that the "sidebar" went away in the mobile view that definitely indicated to me that it was secondary. But yeah, it's also not always a sidebar as it goes to the top in the mobile view. @tyong any thoughts on what we should be naming these things?

jeresig commented:
@matchu Just a heads-up that Tabitha is going to be reviewing the modals today, as well.

matchu commented on packages/wonder-blocks-modal/components/two-column-modal.js:
```diff
@@ -35,33 +42,73 @@ type Props = {
+                <View style={styles.sidebar}>{sidebar}</View>
+            </FlexCell>
+        );
+
```
I wonder if maybe we could pull all this stuff into a separate `_renderSidebar` method? Then use `if` blocks and `return` statements to handle the `null`, `smallSidebar`, and `!smallSidebar` cases? In this form, it's hard for me to keep track of the flow of rendering.

jeresig replied:
Good call! I can do that.

matchu commented on packages/wonder-blocks-modal/components/two-column-modal.js:
```diff
@@ -35,33 +42,73 @@ type Props = {
+                                style={[
+                                    styles.contents,
+                                    smallSidebar && styles.contentsSmall,
+                                    !!footer && styles.contentsHasFooter,
```
Surprised to see a double-bang here when the line above doesn't have one. Is that a Flow typing thing?

jeresig replied:
Oh - `smallSidebar` is a boolean and `footer` is potentially a node, or null, or whatever. I like trying to keep it clean so that I'm consistently using booleans - don't think Flow cares about this, TBH.

matchu commented on packages/wonder-blocks-modal/components/two-column-modal.js:
```diff
@@ -35,33 +42,73 @@ type Props = {
-        display: "flex",
-        flexDirection: "row",
-        alignItems: "stretch",
+        margin: "0 auto",
```
Curious what this self-centering rule is for.

jeresig replied:
I did this to center the modal inside the container. Although maybe I shouldn't be doing that and should leave it up to the container to take care of? Thoughts?

matchu replied:
Yeah, shrug! I would generally expect the modal to handle alignment itself, but this isn't necessarily a ridiculous default. (Then again, I'm not sure how easy it is for parents to override things like `margin: 0 auto`—if it's tricky, maybe it's worth erring toward giving the parent more freedom.)

matchu commented on packages/wonder-blocks-modal/components/two-column-modal.js:
```diff
@@ -78,32 +125,85 @@ const styles = StyleSheet.create({
-        boxSizing: "border-box",
-        flex: "0 0 50%",
+    containerNoSidebar: {
+        minHeight: "auto",
```
What does this rule do?

jeresig replied:
That was overriding the minHeight defined in `container`, but I think I'm going to try and reformulate this so it isn't necessary.

matchu commented on packages/wonder-blocks-modal/components/two-column-modal.js:
```diff
@@ -78,32 +125,85 @@ const styles = StyleSheet.create({
         paddingTop: 64,
         paddingBottom: 64,
+        minHeight: 464,
+        flex: "1 1 50%",
```
I feel like this rule is a bit misleading, because the 50% flex-basis isn't really relevant for the small-sidebar and 1-column cases, but happens to work anyway due to clever subtleties. I'd kinda prefer that we break out the styles for these various cases into entirely different style classes, to help make the expected behavior clearer for future readers.

jeresig replied:
Hmm, ok! I'll see if I can split apart some of these rules to make them clearer (rather than having styles that override each other).

matchu commented on packages/wonder-blocks-modal/components/two-column-modal.js:
```diff
@@ -78,32 +125,85 @@ const styles = StyleSheet.create({
+    },
+
+    contentsSmall: {
+        margin: "0 auto",
```
What does this do?

jeresig replied:
In the "2-column with subordinated preview" (or in the implementation, when `smallSidebar` is `true`) the contents are essentially centered in the container - although I realize that I got the margins slightly wrong (as you noted where it dropped to 0) so I'll fix that.

jeresig commented:
@matchu I think I addressed most of your feedback! I also split out all of the Grid changes into a separate diff to make it easier to review. Also I made a new OneColumnGrid component and broke that out and that helped to simplify things. I also made a new private ModalFooter component which is shared between the One and Two-column grid. Would love a re-review, thank you!

matchu commented on packages/wonder-blocks-modal/components/one-column-modal.md:
```diff
@@ -0,0 +1,119 @@
+const OneColumnModal = require("./one-column-modal.js").default;
+
+const styles = StyleSheet.create({
+    example: {
```
I updated some of these styles in #64, so we'll probably want to keep that in mind when we merge these changes together. (Maybe it's approaching time to actually pull them into a shared example component!)

jeresig replied:
Good idea! I'll wait on you and I can go back through and update these once landed.

matchu commented on packages/wonder-blocks-modal/components/two-column-modal.js:
```diff
@@ -27,49 +29,94 @@ type Props = {
+    /** The main column's content. */
+    content: React.Node,
+
+    /** Should the sidebar be made smaller? (e.g. 1 column wide) */
```
nit: should this be an "i.e." ("in other words") instead of an "e.g." ("for example")? Or are we maybe using that to suggest that being exactly 1 column isn't part of the contract, so we're framing it as an example of what we might do?

jeresig replied:
Right - I'll fix this, I think it should probably just be `(exactly 1 column wide)`

matchu commented on packages/wonder-blocks-modal/components/two-column-modal.js:
```diff
@@ -79,31 +126,63 @@ const styles = StyleSheet.create({
+    },
+
+    footer: {
+        position: "absolute",
```
Hmm gotcha, I think I see why depending on pure Flexbox is trickier in the two-column case than the one-column case: the grid component has padding, but we want the footer to escape that padding.

I'm still surprised that absolute positioning doesn't disrupt the Flex children here, but I guess, if it works, it works! Long-term, I wonder if this is a hint that we'll need a built-in escape hatch to allow certain grid children to be full-bleed, but I'm okay with this for the modal case if it's working consistently!

jeresig replied:
Yeah, we're definitely going to want an "escape hatch"/"full-bleed" component. For now I just have the footer that I've hacked together but I'm not sure if it makes sense to use absolute positioning for this stuff all the time? I can think about this more and maybe I can come up with a better solution in a follow-up commit.

jeresig commented on packages/wonder-blocks-modal/components/one-column-modal.js:
```diff
@@ -0,0 +1,119 @@
+    },
+
+    footer: {
+        position: "absolute",
```
Yeah, this is to escape the padding - and unfortunately it's not possible to move it outside (I would've loved to have done this!) as its positioning is also defined by the grid system 😭

jeresig commented:
@matchu I'm investigating a fix for the scrollbar being too close - it definitely has to do with the grid but I'm seeing if I can come up with another way...

matchu commented:
Oh, wait, does this depend on #65 which isn't merged? I think I see now, sorry.

jeresig commented:
@matchu yeahhh - and now that we're not using the grid I'm likely going to start over. I think I'll wait until you've had a go at things though. Feel free to do whatever work you need to do today and I can pick this back up again on Monday.

matchu commented:
Okay, sorry it didn't work out! Thanks John!
